### ISSUE TITLE
Feature request: Support rootless docker

### ISSUE THREAD

**Issue by tehplague** [ddev/ddev#2899]
**Describe the bug**
ddev fails running the testddev project when running under Rootless Docker on Ubuntu. This possibly happens because ddev is unable to determine the exposed ports of the acting containers. At least the syntax errors in the generated docker-compose.yml suggest that.

**To Reproduce**
Install Rootless Docker on Ubuntu (tested with 20.04 and 20.10). Then invoke either
`rm -rf ~/tmp/testddev && ddev poweroff && mkdir ~/tmp/testddev && cd ~/tmp/testddev && ddev config --project-type=php && ddev start`
or use the supplied `test_ddev.sh` script

**Expected behavior**
ddev should start the demo project and work as intended. The project should be accessible through the started webserver.

**Screenshots**
Diagnostic output: https://gist.github.com/tehplague/33b0c7730b997c9a495183ff8795a810

**Version and configuration information (please complete the following information):**

- Host computer OS and Version: Ubuntu 20.04/20.10
- ddev version information: s. above gist
- config.yaml contents for the misbehaving project: not modified
- Do you have any custom configuration (nginx, php, mysql) in the .ddev folder? No.

**Additional context**
Rootless Docker does not provide a `docker0` interface because all proxying happens in userspace. Therefore the usual logic inside ddev (mainly https://github.com/drud/ddev/blob/master/scripts/host-docker-internal.sh#L14) won't work.

**Comment by rfay** [comment]
ddev has never claimed to support rootless docker. If you want to make the case for a feature request, you're welcome to. 

Please remember that ddev works the same across several different platforms, and that's a very high priority.

**Comment by tehplague** [comment]
Hm. The system requirements page (https://ddev.readthedocs.io/en/stable/#system-requirements) does not explicitly define this. Rootless mode can basically be seen as just another mode of running the same Docker packages as if I installed it system-wide.
If you currently do not support this, you should clarify how people have to set up Docker in order to stay compatible with ddev. As for now, this is undefined.

**Comment by rfay** [comment]
A PR or other exploration is welcome. You could experiment by using docker-compose the the generated .ddev/..ddev-docker-compose-full.yaml . I'm sure the issue you found is just the tip of the iceberg and there will be many more.  I'm quite sure that the derivation of host.docker.internal will have to change, and probably lots of other things.

I think the [Docker installation](https://ddev.readthedocs.io/en/latest/users/docker_installation/#linux-installation-docker-and-docker-compose) section of the manual is pretty clear about the expected setup for docker on linux. 

Currently this feature isn't on the roadmap and hasn't been prioritized at all, but if you lead the charge and experiment with it, that could make it happen.

**Comment by tehplague** [comment]
Yes, most likely. `host.docker.internal` only exists on macOS and WSL because it's a hack for the underlying Linux VM which obviously is not needed on Linux. Under rootless Docker each `rootlesskit` instance receives its own `veth` interface and hence it's own "127.0.0.1". Hence, the container won't be able to figure out which port its exposed ports are mapped to.
We definitely have to find a crossplatform solution for that because it's not uncommon for containerized services to access their own exposed ports (e.g. a containerized script that tries to fetch it's own URLs for indexing purposes and so on).

**Comment by rfay** [comment]
No, they've added an official way to get host.docker.internal now in Docker 20.10+, and it obsoletes the way ddev does it (looking up the address using the docker0 interface). So that's one thing that can be done is to detect docker 20.10+ and use the official way when found. You can test that out as well. The work was done in https://github.com/docker/for-linux/issues/264 and it made it into 20.10.  https://github.com/docker/for-linux/issues/264#issuecomment-784985736

**Comment by rfay** [comment]
I think this will be interesting, but this issue hasn't gotten any traction, so closing for now. Happy to reopen if people are interested in it.

**Comment by finalhow** [comment]
@rfay I realize this issue has been closed, but I feel like it would be a useful feature.  As an organization we need to run docker rootless in order to ensure strict security on our systems. 

Also, while this is not supported at the moment, do you have any idea if we could modify the volume entry in the ~/.ddev/.router-compose-full.yaml file and somehow rebuild the ddev-router container?

```
    volumes:
    - type: bind
      source: /var/run/docker.sock
      target: /tmp/docker.sock
      read_only: true
      bind:
        create_host_path: true
```

It seems as though this could solve the issue, at least temporarily.  

Thanks!

**Comment by rfay** [comment]
`~/.ddev/router-compose.*.yaml` is supported, why don't you give it a try? 

I'm not clear on how this solves your problem, but I do think ddev-router might work OK with read-only if that works for you. And that would be a security win.

**Comment by finalhow** [comment]
Thanks for your quick reply.
I had a look at your code and I see a couple of things that may help explain my issue... I think this is where my problem is:

```
./pkg/ddevapp/router_compose_template.yaml:11:      - /var/run/docker.sock:/tmp/docker.sock:ro
```

The path to the docker.sock is hardcoded as /var/run/docker.sock, but in some users case (especially if you run rootless) that path will vary.  Typically the path will be stored in the $DOCKER_HOST env var, which I see you refer to in:

```
./containers/ddev-router/Dockerfile:7:ENV DOCKER_HOST unix:///tmp/docker.sock
```

So, I'm not sure why the Dockerfile has the correct way of getting that socket location, but the router compose one doesn't.  Is that something that could be changed or made to be configurable?

Thanks!

**Comment by finalhow** [comment]
> `~/.ddev/router-compose.*.yaml` is supported, why don't you give it a try?
> 
> I'm not clear on how this solves your problem, but I do think ddev-router might work OK with read-only if that works for you. And that would be a security win.

This would be helpful if I could modify and set the volume mount to the socket path that I need, but I'm not sure how to do this.  What filename would I use here, and how would I rebuild the ddev-router container with this file?

**Comment by rfay** [comment]
Glad it's working for you! If you think this can be done with a nonbreaking change to DDEV a PR would be great.

**Comment by andypost** [comment]
It's interesting because just a few operations needs root

**Comment by rfay** [comment]
The real reason this hasn't been pursued in general is not technical, it's just that it hasn't had wide traction, doesn't matter that much to most people. And it's "another thing".

**Comment by donquixote** [comment]
From my own experience, the reason that for a long time I did run docker as root is because this is what I got out of the box, and I was tired of reading more manual pages. I always wished for docker to not be running as root, but it was not even sure where to start to achieve it.

I imagine that others are in the same situation: They would prefer rootless, but it just seems too much effort.

Now a few days ago I did the switch.
One thing I did have to fix on regular docker projects was the ip address for xdebug. `client_host=192.168.1.2` (host ip address)

Now I notice that ddev does not work anymore.

I imagine if ddev documentation came with an alternative instruction set of "run with rootless docker", that would include the setup of rootless docker from scratch, then people would actually use it.

**Comment by rfay** [comment]
Just so you know @donquixote - DDEV runs its (core) containers as the same user as used on the host, not as root. That's intended to prevent misuse of root. Also the Traefik-based router no longer mounts the docker socket into the contianer, which closes another major security issue.

**Comment by rfay** [comment]
Use `docker context` to change contexts if you have one rootless and one non. Use `docker context ls` to see what contexts are there. DDEV is focused primarily on developer experience, and I don't think we could have that good experience (especially running containers as the user) with rootless. I'm certainly open to exploration and contributions that work on all platforms.

**Comment by donquixote** [comment]
Thanks, this works quite well!
Now I have both running and visit my ddev projects on rootful and the other projects on rootless without having to restart or change anything.

For anyone else reading this:
Run `docker context use rootless` or `docker context use default` to switch the active context.
Run `docker context ls` to see all contexts. The active one has asterisk.
You can also run `docker-compose --context rootless ...` or `docker-compose --context default ...`.

I did not find a `--context` cli option in ddev.

**Comment by rfay** [comment]
DDEV uses the currently selected context.

**Comment by rfay** [comment]
I think this would be a great feature for DDEV, reopening. Sounds like it's pretty easy.

**Comment by larsskj** [comment]
Let me add that Docker has become quite an expensive option in an enterprise environment.

So, where I work, we're abandoning Docker in favor of Podman as fast as possible: It's more secure - and it's free for all. No cumbersome and expensive licensing.

I'm using Podman for all my PHP development, so unfortunately DDEV is out of reach for me as long as it depends on Docker.

**Comment by rfay** [comment]
No, there's a free docker server for every OS. Not just a docker client.  I'm quite sure your company knows how to forbid specific commercial products in its network and knows the difference between Docker Desktop and all the open-source alternatives. And, to note... Mirantis is not involved with any of these.

**Comment by mark-dropbear** [comment]
Just wanted to say that this issue was a blocker for me using `ddev` in managed development environments like [Firebase Studio](https://firebase.google.com/docs/studio) which **only** support rootless docker.

**Comment by rfay** [comment]
Thanks @mark-dropbear - if you're interested in pursuing this and exploring the territory your exploration is welcome. Note that anything that is done has to work for all platforms (or at least be no impact on other platforms.)

**Comment by andypost** [comment]
@rfay quickly checked it [with grok](https://x.com/i/grok/share/XGk1aVMGGDrxyEkCJ9jZLfFCr) and it said the only change required is a way to detect host's IP

**Comment by rfay** [comment]
Please explore!

**Comment by stasadev** [comment]
I tested rootless Docker with DDEV v1.24.7 on Ubuntu 24.04:

- I completed all prerequisites: https://docs.docker.com/engine/security/rootless/#prerequisites
- Installation with `dockerd-rootless-setuptool.sh install` was straightforward: https://docs.docker.com/engine/security/rootless/#install
- Encountered an issue on `ddev start` with privileged ports 80/443, resolved with: https://docs.docker.com/engine/security/rootless/#exposing-privileged-ports
- But there are ownership issues: `/var/www/html` inside the container is owned by `root:root`:

	```
	stas@ubuntu2404:~/workspace/l12$ docker context ls
	NAME         DESCRIPTION                               DOCKER ENDPOINT                     ERROR
	default      Current DOCKER_HOST based configuration   unix:///var/run/docker.sock         
	rootless *   Rootless mode                             unix:///run/user/1000/docker.sock
	
	stas@ubuntu2404:~/workspace/l12$ ddev start
	(all good)
	
	stas@ubuntu2404:~/workspace/l12$ ddev exec ls -ld /var/www/html
	drwxrwxr-x 4 root root 4096 Aug 12 16:48 /var/www/html
	
	stas@ubuntu2404:~/workspace/l12$ ddev exec touch dummy.php
	touch: cannot touch 'dummy.php': Permission denied
	Failed to execute command `touch dummy.php`: exit status 1
	
	stas@ubuntu2404:~/workspace/l12$ ddev exec whoami
	stas
	
	stas@ubuntu2404:~/workspace/l12$ whoami
	stas
	```
- Ownership can be "solved" with Mutagen:

    ```
    ddev poweroff
    ddev config global --performance-mode=mutagen
    ```

**Comment by stasadev** [comment]
If anyone wants to help with testing, review, or share your thoughts, please check:

- https://github.com/ddev/ddev/pull/7702#issuecomment-3378429468
- https://pr-476.ddev-com-fork-previews.pages.dev/blog/podman-and-docker-rootless/ (https://github.com/ddev/ddev.com/pull/476)

**Comment by stasadev** [comment]
Available in [DDEV HEAD](https://docs.ddev.com/en/stable/developers/building-contributing/#testing-latest-commits-on-head). General availability is planned for DDEV v1.25.0.
